### PR TITLE
Refactor object locking for UserSessions

### LINKED ISSUE #12717: Make locking support for map storage specific to logical entities
### Description

The PR #12241 introduced `LockObjectsForModification` for the logical layer as a hint for the stores to be able to lock entities to avoid situation of concurrent modification in multiple concurrent sessions.

During the review it was round that the method `lockObjectsForModification()` doesn't reveal to the storage layer which entities the caller is interested. Therefore a more specific method like `lockUsersSessionsForModification()` should be available. 

In addition to that: the "Single Use Objects Map JPA implementation" would also benefit from locking entities, to avoid race conditions on single-use.

### Discussion

https://github.com/keycloak/keycloak/pull/12241#discussion_r905782144

### Motivation

Fine grained locking that on the other hand doesn't lock too much that would prevent parallel processing.

### Details

_No response_

### PR DESCRIPTION
**Pull request by ahus1**
Closes #12717

Any additional use of this - for example in Single Use Objects Map JPA implementation - will be postponed as it is about to be used for CockroachDB first where test might fail without locking.

### CODE REVIEW

**Comment by ahus1** [comment]
@vramik - this generalizes the object level locking also for other model types. 

The parent issue mentions to apply this to single use objects; I decided to not put it in here due to time constraints. It will be split of as a separate issue.

cc: @sguilhen - You might want to watch this PR as it might be the base to solve some of the concurrency problems with CockroachDB - that's also the reason I'd rather wanted to have this prepared earlier than to include also the single use objects in it.

**Review by vramik — DISMISSED** [review]
Thank you @ahus1 for the PR. LGTM and it looks much cleaner in the code. I have two minor suggestions regarding naming, feel free to ignore them, it's more like nitpicks from my POV.

**Comment by sguilhen** [comment]
@ahus1 Thanks for the heads up. I've finished collecting/organizing the CRDB failures (see #11618) so it should be easy for anyone to reproduce the issues and test possible fixes.

File: server-spi-private/src/main/java/org/keycloak/utils/LockObjectsForModification.java [hunk 1]
```diff
@@ -38,21 +42,36 @@ public class LockObjectsForModification {
...
 
-    public static boolean isEnabled(KeycloakSession session) {
-        return session.getAttribute(ATTRIBUTE) != null;
+    public static <V> V lockUserSessionsForModification(KeycloakSession session, CallableWithoutThrowingAnException<V> callable) {
+        return lockObjectsForModification(session, UserSessionModel.class, callable);
     }
 
-    public static <V> V lockObjectsForModification(KeycloakSession session, CallableWithoutThrowingAnException<V> callable) {
-        if (LockObjectsForModification.isEnabled(session)) {
+    private static <V> V lockObjectsForModification(KeycloakSession session, Class<?> model, CallableWithoutThrowingAnException<V> callable) {
```
Comment (sguilhen) [inline server-spi-private/src/main/java/org/keycloak/utils/LockObjectsForModification.java:68]: Whats the idea behind making this private? If we need to lock other areas later (e.g. auth sessions) we would add an area-specific method here, like `lockAuthSessionsForModification(session, callable)`?

Comment (ahus1) [reply server-spi-private/src/main/java/org/keycloak/utils/LockObjectsForModification.java:68]: yes, that was the intention - add area specific methods as needed. 

I don't want people to pick the classes manually; they won't be type safe so they could place any class there. There is no base interface/class for the models AFAIK.

**Review by vramik — APPROVED** [review]

**Review by pedroigor — APPROVED** [review]

**Pull request merged**